src/server/saved_objects/client/saved_objects_client.test.js

```diff
@@ -1,643 +1,96 @@
-    it('can filter by fields', async () => {
-      await savedObjectsClient.find({ fields: ['title'] });
-
-      sinon.assert.calledOnce(callAdminCluster);
-      sinon.assert.calledWithExactly(callAdminCluster, sinon.match.string, sinon.match({
-        _source: [
-          '*.title', 'type', 'title'
-        ]
-      }));
-
-      sinon.assert.notCalled(onBeforeWrite);
-    });
-  });
-
-  describe('#get', () => {
-    beforeEach(() => {
-      callAdminCluster.returns(Promise.resolve({
-        _id: 'index-pattern:logstash-*',
-        _type: 'doc',
-        _version: 2,
-        _source: {
-          type: 'index-pattern',
-          ...mockTimestampFields,
-          'index-pattern': {
-            title: 'Testing'
-          }
-        }
-      }));
-    });
-
-    it('formats Elasticsearch response', async () => {
-      const response = await savedObjectsClient.get('index-pattern', 'logstash-*');
-      sinon.assert.notCalled(onBeforeWrite);
-      expect(response).toEqual({
-        id: 'logstash-*',
-        type: 'index-pattern',
-        updated_at: mockTimestamp,
-        version: 2,
-        attributes: {
-          title: 'Testing'
-        }
-      });
-    });
-
-    it('prepends type to the id', async () => {
-      await savedObjectsClient.get('index-pattern', 'logstash-*');
-
-      sinon.assert.notCalled(onBeforeWrite);
-      sinon.assert.calledOnce(callAdminCluster);
-      sinon.assert.calledWithExactly(callAdminCluster, sinon.match.string, sinon.match({
-        id: 'index-pattern:logstash-*',
-        type: 'doc'
-      }));
-    });
-  });
-
-  describe('#bulkGet', () => {
-    it('accepts a array of mixed type and ids', async () => {
-      callAdminCluster.returns({ docs: [] });
-
-      await savedObjectsClient.bulkGet([
-        { id: 'one', type: 'config' },
-        { id: 'two', type: 'index-pattern' }
-      ]);
+  SavedObjectsRepository.mockClear();
+  SavedObjectsRepository.mockImplementation(() => mockSavedObjectsRepository);
+});
 
-      sinon.assert.calledOnce(callAdminCluster);
-      sinon.assert.calledWithExactly(callAdminCluster, sinon.match.string, sinon.match({
-        body: {
-          docs: [
-            { _type: 'doc', _id: 'config:one' },
-            { _type: 'doc', _id: 'index-pattern:two' }
-          ]
-        }
-      }));
+test(`#constructor`, () => {
+  const options = {};
+  new SavedObjectsClient(options);
+  expect(SavedObjectsRepository).toHaveBeenCalledWith(options);
+});
 
-      sinon.assert.notCalled(onBeforeWrite);
-    });
+test(`#create`, () => {
+  const client = new SavedObjectsClient();
 
-    it('returns early for empty objects argument', async () => {
-      callAdminCluster.returns({ docs: [] });
+  const type = 'foo';
+  const attributes = {};
+  const options = {};
+  client.create(type, attributes, options);
 
-      const response = await savedObjectsClient.bulkGet([]);
+  expect(mockSavedObjectsRepository.create).toHaveBeenCalledWith(type, attributes, options);
```

Comment:
All of these tests need to verify that they return the same reference that the repository function returns as well, or whatever the expectations are there.

Reply:
They should probably verify that repository exceptions are passed through as well.

Reply:
I agree with regard to the returns being passed through, but it feels excessive to test that errors propagate.

Reply:
Why's that? Error handling is a critical part of any public API, and it's going untested here.

Don't get me wrong, it is true that the old tests weren't testing any error states, which while that's not good, it's not something we need to block this PR on. However, what you're doing here is to effectively treat the repository behaviors as the source of truth for the corresponding public client functions. So if I go make some change to how errors are handled in the repository and at that time I update the repository tests to reflect that handling, there is no testing in place to verify that I didn't just break the client API.

Reply:
So, just to be clear, you're recommending that we duplicate the tests in the repository for the specific errors being thrown to be tests for the client?

Reply:
We talked this over, and since the repository functions are async, verifying the returned value should handle this as well.